src/vis_cpu/vis_cpu.py

```diff
@@ -20,66 +23,142 @@ def profile(fnc):
         return fnc
 
 
-def construct_pixel_beam_spline(bm_cube):
-    """Construct bivariate spline for pixelated beams for all antennas.
+logger = logging.getLogger(__name__)
 
-    Uses the ``scipy.interpolate.RectBivariateSpline`` function.
 
-    Parameters
-    ----------
-    bm_cube : array_like
-        Pixelized beam maps for each antenna (must be real-valued).
-        Shape: (NANT, BEAM_PIX, BEAM_PIX) if unpolarized, or
-        (NANT, NAXES, NFEEDS, BEAM_PIX, BEAM_PIX) if polarized.
+def _wrangle_beams(beam_idx, beam_list, polarized, nant, freq, interp=True):
+    """Perform all the operations and checks on the beams."""
+    # Get the number of unique beams
+    nbeam = len(beam_list)
 
-    Returns
-    -------
-    splines : list of fn
-        List of interpolation functions, one for each antenna, in order.
-    """
-    # Raise error if complex
-    if np.any(np.iscomplex(bm_cube)):
-        raise TypeError(
-            "bm_cube cannot be complex. Interpolate real and "
-            "imaginary components separately."
+    # Check the beam indices
+    if beam_idx is None:
+        if nbeam == 1:
+            beam_idx = np.zeros(nant, dtype=int)
+        elif nbeam == nant:
+            beam_idx = np.arange(nant, dtype=int)
+        else:
+            raise ValueError(
+                "If number of beams provided is not 1 or nant, beam_idx must be provided."
+            )
+    else:
+        assert beam_idx.shape == (nant,), "beam_idx must be length nant"
+        assert all(
+            0 <= i < nbeam for i in beam_idx
+        ), "beam_idx contains indices greater than the number of beams"
+
+    if interp:
+        # make sure we interpolate to the right frequency first.
+        beam_list = [
+            bm.interp(freq_array=np.array([freq]), new_object=True, run_check=False)
+            if isinstance(bm, UVBeam)
+            else bm
+            for bm in beam_list
+        ]
+
+    if polarized and any(b.beam_type != "efield" for b in beam_list):
+        raise ValueError("beam type must be efield if using polarized=True")
+    elif not polarized and any(
+        (
+            b.beam_type != "power"
+            or getattr(b, "Npols", 1) > 1
+            or b.polarization_array[0] not in [-5, -6]
+        )
+        for b in beam_list
+    ):
+        raise ValueError(
+            "beam type must be power and have only one pol (either xx or yy) if polarized=False"
+        )
+
+    return beam_list, nbeam, beam_idx
+
+
+def _evaluate_beam_cpu(
+    beam_list,
+    tx,
+    ty,
+    polarized,
+    nbeam,
+    nax,
+    nfeed,
+    freq,
+    nsrcs_up,
+    complex_dtype,
+    spline_opts=None,
+):
```

Comment:
Would also be useful to document the arguments here.

Reply:
Yes. Done.